src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
-                _sets = new HashSet<HttpConnection>[s_numSlots];
-                for (int i = 0; i < _sets.Length; i++)
+                // Get the current time.  This is compared against each connection's last returned
+                // time to determine whether a connection is too old and should be closed.
+                DateTimeOffset now = DateTimeOffset.Now;
+
+                // Find the first item which needs to be removed.
+                int freeIndex = 0;
+                while (freeIndex < list.Count && list[freeIndex].IsUsable(now))
+                {
+                    freeIndex++;
+                }
+
+                // If freeIndex == list.Count, nothing needs to be removed.
+                // But if it's < list.Count, at least one connection needs to be purged.
+                if (freeIndex < list.Count)
                 {
-                    _sets[i] = new HashSet<HttpConnection>();
+                    // We know the connection at freeIndex is unusable, so dispose of it.
+                    list[freeIndex]._connection.Dispose();
+
+                    // Find the first item after the one to be removed that should be kept.
+                    int current = freeIndex + 1;
+                    while (current < list.Count)
+                    {
+                        // Look for the first item to be kept.  Along the way, any
+                        // that shouldn't be kept are disposed of.
+                        while (current < list.Count && !list[current].IsUsable(now))
+                        {
+                            list[current]._connection.Dispose();
+                            current++;
+                        }
+
+                        // If we found something to keep, copy it down to the known free slot.
+                        if (current < list.Count)
+                        {
+                            // copy item to the free slot
+                            list[freeIndex++] = list[current++];
+                        }
+
+                        // Keep going until there are no more good items.
+                    }
+
+                    // At this point, everything at and after freeIndex are garbage connections
+                    // that have been disposed.  Remove them all from the list.
+                    list.RemoveRange(freeIndex, list.Count - freeIndex);
+
+                    // If there are now no connections associated with this pool, we can dispose of it.  Note that 
+                    // the value could be NewPoolSentinel, in which case we don't want to clean it up, as the first
+                    // connection it was created for hasn't yet been added. We also avoid aggressively cleaning up
+                    // pools that have recently been used but currently aren't; if a pool was used since the last
+                    // time we cleaned up, give it another chance.
+                    if (_associatedConnectionCount == 0 && !_usedSinceLastCleanup)
```

Comment:
If I understand correctly, the _associatedConnectionCount is modified by the connection.Dispose call.  Is that correct?  And we are relying on lock reentry to allow that update to happen?  That seems a little scary to me, and at least deserves a comment or two (one above at the Dispose call here, one at HttpConnection.Dispose, to indicate it may be called due to connection timeout and so we should be careful not to do anything weird here.

It might be better to build up a separate list of connectionsToRemove and then dispose them outside of the lock.  That seems a bit safer to me, and would reduce the time we're holding the lock for also.